Benchmarking with ASV

jpn-- opened the pull request:
work on #439

jpn-- committed:
reindex vs merge

jpn-- committed:
start work on benchmarking

jpn-- committed:
as a suite

jpn-- committed:
timing and setup

jpn-- committed:
streamlining

jpn-- committed:
feat: add optional sha256 to example manifest

this can prevent upcoming benchmarking tool from pounding relentlessly on activitysim_resources repo, by caching the large file downloads

jpn-- committed:
download files in setup_cache

jpn-- committed:
finish suite

jpn-- committed:
ex press

jpn-- committed:
Merge branch 'chunkless' into performance1

jpn-- committed:
component logging

jpn-- committed:
more stable logging

jpn-- committed:
no locutor for school location in benchmarking

jpn-- committed:
read-only pipeline

jpn-- committed:
don't write to checkpoints when reading them

jpn-- committed:
don't rollover for every timing

jpn-- committed:
log exceptions in benchmarks

jpn-- committed:
table cleaning for mand tours freq

jpn-- committed:
clean sweep all tables

jpn-- committed:
benchmark cli command

jpn-- committed:
benchmark test

jpn-- committed:
better cli wrapper for asv

jpn-- committed:
locally manage json

jpn-- committed:
better workspace

jpn-- committed:
benchmarks are in the package

jpn-- committed:
local dir

jpn-- committed:
ASIM_ASV_WORKSPACE

jpn-- committed:
fix dir for download

jpn-- committed:
fix models dir

jpn-- committed:
single thread flags

jpn-- committed:
allow asv to be missing

jpn-- committed:
fix args

jpn-- committed:
change suite to module

jpn-- committed:
no self

jpn-- committed:
partial

jpn-- committed:
discoverable

jpn-- committed:
remote intermediate funcs

jpn-- committed:
call the partial

jpn-- committed:
more parts

jpn-- committed:
reraise

jpn-- committed:
refactor

jpn-- committed:
names

jpn-- committed:
to tour mode

jpn-- committed:
tinker names

jpn-- committed:
just pretty

jpn-- committed:
generic cleaning

jpn-- committed:
Merge branch 'condaforge-dev' into performance1

jpn-- committed:
define mtc1 as 20K households

jpn-- committed:
ready for win

jpn-- committed:
pb branches

jpn-- committed:
Merge branch 'faster-maz-maz' into performance2

jpn-- committed:
Merge branch 'speeding-up-chunking' of https://github.com/ActivitySim/activitysim into performance2

jpn-- committed:
more aggressive chunkless

jpn-- committed:
asv commit logging

bstabler commented:
@jpn-- please PR to develop when ready

jpn-- committed:
Merge branch 'master' into performance2

jpn-- committed:
Merge branch 'checksum-downloads' into performance2

# Conflicts:
#	activitysim/cli/create.py

jpn-- committed:
asv sandag1

jpn-- committed:
setup_cache as func

jpn-- committed:
fix sandag1 benchmarks

jpn-- committed:
allow other configs dirs in prerun

jpn-- committed:
data and output definable

jpn-- committed:
checksums were correct

jpn-- committed:
dirs in component setup

jpn-- committed:
only use 20K hh for sandag tests

jpn-- committed:
documentation updates

jpn-- committed:
sandag2

jpn-- committed:
submitting benchmarks

jpn-- committed:
fix sandag2 benchmarks

jpn-- committed:
note about escapes

jpn-- committed:
pycodestyle

jpn-- committed:
sandag3

jpn-- committed:
extra cli run args

jpn-- committed:
extra logging

jpn-- committed:
performanceTest

jpn-- committed:
elapsed time logging

jpn-- committed:
elapsed time logging

jpn-- committed:
use cached accessibility

jpn-- committed:
settings remap

jpn-- committed:
Merge commit '6ba7e8ff386c95e3bf1b05864cc42461f1e953aa' into performanceTest

jpn-- committed:
print line

jpn-- committed:
profiling

jpn-- committed:
fix branch names

jpn-- committed:
use smaller sandag 2-zone

jpn-- committed:
mp complete

jpn-- committed:
profile inspector

jpn-- committed:
sandag2 mp complete

jpn-- committed:
Merge commit 'fd077d644fdb41200031f2860cfc0d60e2b90574' into performanceTest

jpn-- committed:
pipeline hash for benchmarking

jpn-- committed:
absolute import

jpn-- committed:
manifest points to a place

jpn-- committed:
allow remote manifest files to be gzipped, then expanded after download.

jpn-- committed:
harmonize sandag3

jpn-- committed:
sandag1

jpn-- committed:
commit token

jpn-- committed:
extra logging, stable option for prerun

jpn-- committed:
remove pipeline hash

jpn-- committed:
sandag1e

jpn-- committed:
mostly templated sandag

jpn-- committed:
cleanup

jpn-- committed:
sandag3f

jpn-- committed:
use full population in data for MP complete test

jpn-- committed:
benchmark latest

jpn-- committed:
benchmarking configs

jpn-- committed:
allow missing models

jpn-- committed:
sandag2f

jpn-- committed:
batch

jpn-- committed:
update mp_complete name

jpn-- committed:
rename to fuller names

jpn-- committed:
skip a few

jpn-- committed:
farm out setup cache

jpn-- committed:
template in componentwise

jpn-- committed:
normal logger

jpn-- committed:
cwd error

jpn-- committed:
fix dirs

jpn-- committed:
template mp complete

jpn-- committed:
template complete does not work

jpn-- committed:
one ping only

jpn-- committed:
fix sandag 2e

jpn-- committed:
blacken

jpn-- committed:
cwise template blackened

jpn-- committed:
remove cwise template

jpn-- committed:
fix mtc1full benchmark

jpn-- committed:
clean old benchmark files

jpn-- committed:
hook to fix paths in benchmark output

jpn-- committed:
some docs

jpn-- committed:
documentation

jpn-- committed:
writing new benchmarks doc

jpn-- committed:
explicit benchmark versioning

jpn-- committed:
use std name

jpn-- committed:
consistent single-thread benchmark size

jpn-- committed:
tracking for mp benchmarks

jpn-- committed:
fix log file path

jpn-- committed:
fix track read

jpn-- committed:
clean up

jpn-- committed:
minimally functional multithread benchmark

jpn-- committed:
todo note for MP bench

jpn-- committed:
note about MP benchmarks

jpn-- committed:
pycodestyle

jpn-- committed:
remove config editing

jpn-- committed:
strip ruamel, not needed

jpn-- committed:
fix to use dynamic configs dir

jpn-- committed:
Merge branch 'develop' of https://github.com/ActivitySim/activitysim into performance1

jpn-- committed:
use different branch

jpn-- committed:
Merge commit 'd3ee01edaf17b093c14e559f45a029eb47c2ee5d' into performance1

# Conflicts:
#	activitysim/examples/example_manifest.yaml

jpn-- committed:
benchmarking docs

jpn-- committed:
point to final asim-benchmarks

jpn-- committed:
Merge commit '3df695bd2bf921aa46d296a09889b68087b8c911' into performance1

jpn-- committed:
doc formatting cleanup

jpn-- committed:
asv continuous

jpn-- committed:
include benchmarking in docs table of contents

jpn-- commented:
This work is ready for review, and passes tests in Travis.  The benchmarking results website is online at https://activitysim.github.io/asim-benchmarks/

dhensle commented on docs/benchmarking.rst:
```diff
@@ -0,0 +1,259 @@
+Publishing the standard airspeed velocity content to GitHub pages is a built-in
+feature of the command line tool.  Simply run::
+
+    activitysim benchmark gh-pages
```
All of the above steps were run in order, but this step crashed with the following error:  

    activitysim benchmark gh-pages
    · <== benchmarking activitysim ==>
    ·  workspace: E:\Projects\Clients\activitysim\asv_benchmarking\asim-benchmarks
    ·  activitysim installation: e:\projects\clients\activitysim\asv_benchmarking\activitysim\activitysim
    ·  local git repo available: True
    ·  local git repo: ..\activitysim
    [ 11.11%] · Loading machine info
    [ 22.22%] · Getting params, commits, tags and branches
    [ 22.22%] ·· Error running C:\Users\david.hensle\.conda\envs\ASIM-BENCH\Library\bin\git.exe rev-list --first-parent master (exit status 128)
                 STDOUT -------->
    
                 STDERR -------->
                 fatal: ambiguous argument 'master': unknown revision or path not in the working tree.
                 Use '--' to separate paths from revisions, like this:
                 'git <command> [<revision>...] -- [<file>...]'
    
    Traceback (most recent call last):
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\Scripts\activitysim-script.py", line 33, in <module>
        sys.exit(load_entry_point('activitysim', 'console_scripts', 'activitysim')())
      File "e:\projects\clients\activitysim\asv_benchmarking\activitysim\activitysim\cli\main.py", line 37, in main
        sys.exit(asim.execute())
      File "e:\projects\clients\activitysim\asv_benchmarking\activitysim\activitysim\cli\cli.py", line 27, in execute
        args.afunc(args)
      File "e:\projects\clients\activitysim\asv_benchmarking\activitysim\activitysim\cli\benchmark.py", line 245, in benchmark
        result = args.func(args)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\commands\__init__.py", line 49, in run_from_args
        return cls.run_from_conf_args(conf, args)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\plugins\github.py", line 40, in run_from_conf_args
        return cls.run(conf=conf, no_push=args.no_push, rewrite=args.rewrite)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\plugins\github.py", line 47, in run
        Publish.run(conf)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\commands\publish.py", line 163, in run
        branches = dict(
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\commands\publish.py", line 164, in <genexpr>
        (branch, repo.get_branch_commits(branch))
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\plugins\git.py", line 176, in get_branch_commits
        return self.get_hashes_from_range(self.get_branch_name(branch))
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\plugins\git.py", line 124, in get_hashes_from_range
        output = self._run_git(args, valid_return_codes=(0, 1), dots=False)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\plugins\git.py", line 73, in _run_git
        return util.check_output([self._git] + args, env=env, **kwargs)
      File "C:\Users\david.hensle\.conda\envs\ASIM-BENCH\lib\site-packages\asv\util.py", line 685, in check_output
        raise ProcessError(args, retcode, stdout, stderr)
    asv.util.ProcessError: Command 'C:\Users\david.hensle\.conda\envs\ASIM-BENCH\Library\bin\git.exe rev-list --first-parent master' returned non-zero exit status 128

jpn-- replied:
It looks like perhaps you don't have the 'master' branch tracked in your local git repo.  It appears you need to have that.  Also, I added a note in these docs that the `gh-pages` command is only expected to work when the user has write-access to the `ActivitySim/asim-benchmarks` repository on GitHub.

dhensle commented on docs/benchmarking.rst:
```diff
@@ -0,0 +1,259 @@
+    pre-commit step is required.
+
+Users may find it simpler to skip the last step on the command line, and simply
+visit their fork on GitHub.com to use the web interface to open a pull request.
```
Provide some guidance on when it would be valuable to submit benchmarks.

jpn-- replied:
We as a group need to have some discussions soon as to how to manage and track benchmarks over time, and/or whether we want to centralize the process of benchmarking on a common reference machine.

dhensle commented on docs/benchmarking.rst:
```diff
@@ -0,0 +1,259 @@
+    pre-commit run    # required on Windows only, see note
+    git commit -m "adding benchmark results"
+    git push
+    gh pr create
```
I created a [pull request](https://github.com/ActivitySim/asim-benchmarks/pull/3) for the sandag example running on one of RSG's servers.  Results look fine to me, but you might want to review and ensure following these steps produced what you would expect.

jpn-- replied:
Looks ok, except your local machine paths were in the commit.  I've been trying to figure out how to manage that automatically (unsuccessfully so far).  It's not a big deal so I don't want to block the ASV work because of it.

dhensle commented on docs/benchmarking.rst:
```diff
@@ -0,0 +1,259 @@
+called `airspeed velocity <https://asv.readthedocs.io/en/stable/>`__.
+
+The benchmarking process is closely tied to ActivitySim's *git* repository,
+so it is recommended that you use Git to clone the repository from GitHub.
```
I think some high level information on what Benchmarking does, why its a useful feature, when to use it, etc.  would be helpful to understand the process and have users buy into benchmarking their setups.

jpn-- replied:
Yup.

dhensle commented on docs/benchmarking.rst:
```diff
@@ -0,0 +1,259 @@
+A basic framework for multi-processing benchmarks has been implemented and is
+demonstrated in the ``mtc1mp4`` benchmark file. However, work remains to write
+a stable process to execute chunking training for each machine prior to running
+the production-version benchmarks that will be meaningful for users.
```
I could see developers wanting to check performance of model as incremental updates are made.  Often these updates are made to a model outside of the ActivitySim example folder.  Does an actual ActivitySim example need to be created in order to benchmark an ActivitySim run, or can we just modify the basic template with any path to an ActivitySim run?

jpn-- replied:
As currently implemented, the benchmarking tool only works for models in the published examples.  We can investigate trying to enable this feature for ad hoc external models, but that's outside the scope of the current task.

dhensle commented on activitysim/benchmarking/benchmarks/mtc1full.py:
```diff
@@ -0,0 +1,83 @@
+DYNAMIC_CONFIG_DIR = "bench_configs"
+DATA_DIR = "data"
+OUTPUT_DIR = "output"
+COMPONENT_NAMES = [
```
Is this just the models that we want to benchmark?  Can it be any selection of models?  e.g. I see that initialize_landuse, and initialize_households are missing.  Do we just not care about their times?  Propose that this file's options be included in the documentation with descriptions for each variable that needs to be set.

dhensle commented on activitysim/benchmarking/benchmarks/mtc1mp4.py:
```diff
@@ -0,0 +1,81 @@
+    # but has a bad habit of crashing due to memory allocation errors on
+    # all but the tiniest of examples. It would be great to fix the MP
+    # benchmarks so they use chunking, automatically configure for available
+    # RAM, and run a training-production cycle to get useful timing results.
```
Does this mean that no chunking is allowed and that agencies can't benchmark their full model times?

jpn-- replied:
This is basically correct: ASV as currently implemented is not quite ready to be used to benchmark full-scale multi-process mode tests.  The dynamic chunking process isn't quite automated enough to be portable to a new machine with resources that are not identical to where training occurred; and automatically selecting "good" training settings is not yet available.  Until full-scale models don't require manual tuning for good performance, benchmarking for that purpose will need to be manual also.

dhensle commented on activitysim/benchmarking/componentwise.py:
```diff
@@ -0,0 +1,601 @@
+    settings_file_name=None,
+):
+    """
+    Pre-run the models, checkpointing everything.
```
What does pre-running the models mean and why is it needed?  More documentation here.

dhensle commented on activitysim/benchmarking/componentwise.py:
```diff
@@ -0,0 +1,601 @@
+    config_overload_dir="dynamic_configs",
+):
+    """
+
```
Comment needed on what this method does.  Considering this is one of two primary function calls from the benchmark templates, it would be helpful for any developers looking to make modifications.

dhensle replied:
@jpn-- I think you missed this one

dhensle commented on activitysim/benchmarking/componentwise.py:
```diff
@@ -0,0 +1,601 @@
+    timeout_=36000.0,  # ten hours,
+    version_="1",
+):
+
```
Docstring needed.

dhensle commented on activitysim/benchmarking/componentwise.py:
```diff
@@ -0,0 +1,601 @@
+    try:
+        if config.setting("multiprocess", False):
+            raise NotImplementedError(
+                "multiprocess component benchmarking is not yet implemented"
```
Looks like multiprocessing is implemented later in the file.  Having this error here and the commented out code below is confusing.  Removing the contradictory statements/code would be helpful.

jpn-- replied:
Sorry for the confusion.  I've added a clarifying comment in the code here for future developers (including future me).

Component level timings for multiprocess benchmarking are not generated using this code, which re-runs individual
components inside a timing loop.  Instead, those benchmarks are generated by running the model all the way through once during "setup", and then extracting the per-component timings from logs later.

dhensle commented on activitysim/benchmarking/latest.py:
```diff
@@ -0,0 +1,181 @@
+            "Simply give the file name, which should be a text file "
+            "containing a number of activitysim benchmark commands.",
+            description="Run a set of benchmark suites based on a batch file.",
+        )
```
This is great.  I think this functionality should be included in the main documentation file so people will know about it.

jpn-- replied:
Done.

dhensle commented on activitysim/cli/benchmark.py:
```diff
@@ -0,0 +1,255 @@
+    # "html_dir": "../activitysim-asv/html",
+    # List of branches to benchmark. If not provided, defaults to "master"
+    # (for git) or "default" (for mercurial).
+    "branches": ["master", "develop", "performanceTest"],
```
Does it make sense to have hard coded branches here?

jpn-- replied:
Not at this point.  I've added a commit to change this to get the current branch dynamically.

dhensle commented on activitysim/benchmarking/componentwise.py:
```diff
@@ -0,0 +1,601 @@
+            benchmarking=True,
+            checkpoints=True,
+            trace_hh_id=None,
+            chunk_training_mode="disabled",
```
If we have chunking enabled in our settings file, does it get overwritten here?  If so, this needs to be clearly and obviously stated in the main documentation file that chunking is not implemented.

jpn-- replied:
Added a warning in the documentation: 
    The benchmarks do not currently use ActivitySim's dynamic chunking features,
    as these require manual configuration and training on a per-machine basis
    to ensure good performance.

dhensle reviewed (changes requested):
Pull request in general looks really good.  Code needs to be more consistent when it comes to chunking and multiprocessing (there are multiple spots that seem to contradict each other as to whether multiprocessing & chunking are enabled or not).  More documentation would help developers and users understand the functionality and how to use it appropriately.  I did get an error when trying to publish to GitHub pages that should be addressed.

Does a TravisCI test need to be created to test the benchmarking functionality?

jpn-- committed:
add doc for benchmark batch

thank you @dhensle for the suggestion https://github.com/ActivitySim/activitysim/pull/456/files/5142bd7832f946242b68f475d08d990b24885161#r773383658

jpn-- committed:
dynamically get current branch for asv

jpn-- committed:
some high level information on what Benchmarking does

jpn-- committed:
need to sync to publish

jpn-- committed:
branch arg

jpn-- committed:
add docs

jpn-- committed:
Merge commit '74569ceae1f0f878de75ea328a0a0115abddeb97' into performance1

jpn-- committed:
constrain numpy

temporary fix cf #533

jpn-- committed:
require xarray and sharrow for larch testing

jpn-- commented:
@dhensle, I believe we have addressed all of the items you identified in your review, can you confirm?  Thanks

jpn-- committed:
add docstring for template_setup_cache

dhensle reviewed (approved):
All of my previous comments have been addressed.  Looks good to me!

jpn-- committed:
Merge branch 'develop' into performance1

jpn-- committed:
Update test_larch_estimation.py

fix for dependency changes that result in small parameter estimation variance

jpn-- committed:
pycodestyle
